### DISCUSSION TITLE
Download .so libraries for Android

### DISCUSSION THREAD

**Discussion by drpapp** [FreeRDP/FreeRDP#9051] [General]
Hello,

I tried to built the native FreeRDP libraries for Android with ./scripts/android-build-freerdp.sh following the docs, but got all kind of errors. Is it possible to download the latest .so files from somewhere? It would be a great help overall.
Thanks!

**Comment by akallabeth** [comment]
no download, you need to build it with a proper `NDK`.
if you have issues, then check your development environment is set up properly.

**Reply by drpapp** [reply]
Ok, I'm trying :)

**Reply by akallabeth** [reply]
@drpapp did run a simple `./scripts/android-build-freerdp.sh --ndk ~/Android/Sdk/ndk/25.2.9519653/ --sdk ~/Android/Sdk/` on my machine (with a config for `x86_64` and `aarch64` and it is building fine. (just check your `cmake`in android SDK is recent and you only have one installed)

**Reply by drpapp** [reply]
I always got this error:

ndk-build not found in NDK directory D:/Android/Sdk/ndk/25.2.9519653
assuming ndk-build is in path...

/I'm on windows/

**Reply by drpapp** [reply]
Trying to set up a linux in VMware to give it a try

**Reply by drpapp** [reply]
Hello,

I made progress using ubuntu in vmware. The script started running, however I got this error message:

CMake Error at /home/pv/Downloads/android-sdk/cmake/share/cmake-3.27/Modules/FindPackageHandleStandardArgs.cmake:230 (message):
  Could NOT find ZLIB (missing: ZLIB_INCLUDE_DIR)

I installed zlib, but still the same error.
Can you please help me on this?

Thanks.

**Reply by akallabeth** [reply]
@drpapp ok, how about you just do what is written in the `README`?

* Install `SDK`
* Install `NDK`
* Install `cmake` (from the `SDK`, not from `cmake.org`)
* run freerdp build script with proper paths

**Reply by drpapp** [reply]
@akallabeth thanks for the support!

It could be help to someone, who needs an Android application -README is correct, but some additional details:

I tried to built the libraries on windows, but never succeeded.

1, install any linux OS in Vmware, I used Ubuntu.
2. Install Android Studio
3. once done, open SDK manager, under SDK tool select NDK and CMake (it will download and install these tools)
4. install make: "apt install make"
5. in a terminal run script example: " ./scripts/android-build-freerdp.sh --ndk /home/pv/Android/Sdk/ndk/25.2.9519653/ --sdk /home/pv/Android/Sdk/   "
*change the NDK subdirectory to the current version
6. you will find the libraries under /FreeRDP/client/Android/Studio/freeRDPCore/src/main/jniLibs/

*DO NOT install the required software components alone, like separate android sdk, then cmake from cmake,org etc. It won't work.


Always read through the error msg if the script fails, like below. In this case you will find in the middle, that the error is:

**/home/pv/Downloads/FreeRDP/scripts/android-build-common.sh: line 237: type: make: not found
Command make not found. Install and add it to the PATH.**

You need to install 'make' as it is not in the default Ubuntu distribution. 

pv@pv-virtual-machine:~/Downloads/FreeRDP$ ./scripts/android-build-freerdp.sh --ndk /home/pv/Android/Sdk/ndk/25.2.9519653/ --sdk /home/pv/Android/Sdk/
find: ‘/cmake’: No such file or directory
[RUN] mkdir -p /home/pv/Downloads/FreeRDP/scripts/../client/Android/Studio/freeRDPCore/src/main/jniLibs
[RUN] mkdir -p /home/pv/Downloads/FreeRDP/scripts/../build
[RUN] bash /home/pv/Downloads/FreeRDP/scripts/android-build-cjson.sh --src /home/pv/Downloads/FreeRDP/scripts/../build/cjson --dst /home/pv/Downloads/FreeRDP/scripts/../client/Android/Studio/freeRDPCore/src/main/jniLibs --sdk /home/pv/Android/Sdk/ --ndk /home/pv/Android/Sdk/ndk/25.2.9519653/ --arch armeabi-v7a --target 21 --tag v1.7.15 --hash 5308fd4bd90cef7aa060558514de6a1a4a0819974a26e6ed13973c5f624c24b2
find: ‘/cmake’: No such file or directory
/home/pv/Downloads/FreeRDP/scripts/android-build-common.sh: line 237: type: make: not found
Command make not found. Install and add it to the PATH.
[ERROR] bash /home/pv/Downloads/FreeRDP/scripts/android-build-cjson.sh --src /home/pv/Downloads/FreeRDP/scripts/../build/cjson --dst /home/pv/Downloads/FreeRDP/scripts/../client/Android/Studio/freeRDPCore/src/main/jniLibs --sdk /home/pv/Android/Sdk/ --ndk /home/pv/Android/Sdk/ndk/25.2.9519653/ --arch armeabi-v7a --target 21 --tag v1.7.15 --hash 5308fd4bd90cef7aa060558514de6a1a4a0819974a26e6ed13973c5f624c24b2 retured 1

Good luck!
